crates/sargon-uniffi/src/hierarchical_deterministic/bip32/hd_path_component.rs

```diff
@@ -1,16 +1,822 @@
 use crate::prelude::*;
+use sargon::FromGlobalKeySpace;
+use sargon::FromLocalKeySpace;
 use sargon::HDPathComponent as InternalHDPathComponent;
+use sargon::Hardened as InternalHardened;
+use sargon::SecurifiedU30 as InternalSecurifiedU30;
+use sargon::ToBIP32Str;
+use sargon::Unhardened as InternalUnhardened;
+use sargon::Unsecurified as InternalUnsecurified;
+use sargon::UnsecurifiedHardened as InternalUnsecurifiedHardened;
+use sargon::U30 as InternalU30;
+use sargon::U31 as InternalU31;
 
-pub type HDPathValue = u32;
+use sargon::IsInLocalKeySpace;
+use sargon::IsMappableToGlobalKeySpace;
+use sargon::KeySpace as InternalKeySpace;
 
-#[derive(Clone, PartialEq, Eq, Hash, InternalConversion, uniffi::Record)]
-pub struct HDPathComponent {
-    pub value: HDPathValue,
+/// A discriminator of an `HDPathComponent`.
+#[derive(
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Debug,
+    InternalConversion,
+    uniffi::Enum,
+)]
+pub enum KeySpace {
+    Unsecurified { is_hardened: bool },
+    Securified,
 }
 
 #[uniffi::export]
-pub fn hd_path_component_get_non_hardened_value(
+pub fn new_key_space_sample() -> KeySpace {
+    InternalKeySpace::sample().into()
+}
+
+#[uniffi::export]
+pub fn new_key_space_sample_other() -> KeySpace {
+    InternalKeySpace::sample_other().into()
+}
+
+#[derive(
+    Clone, Debug, PartialEq, Eq, Hash, InternalConversion, uniffi::Enum,
+)]
+pub enum HDPathComponent {
+    Unsecurified(Unsecurified),
+    Securified(SecurifiedU30),
+}
+
+#[uniffi::export]
+pub fn new_hd_path_component_sample() -> HDPathComponent {
+    InternalHDPathComponent::sample().into()
+}
+
+#[uniffi::export]
+pub fn new_hd_path_component_sample_other() -> HDPathComponent {
+    InternalHDPathComponent::sample_other().into()
+}
+
+#[uniffi::export]
+pub fn new_hd_path_component_from_local_key_space(
+    value: u32,
+    key_space: KeySpace,
+) -> Result<HDPathComponent> {
+    InternalHDPathComponent::from_local_key_space(value, key_space.into())
+        .into_result()
+}
+
+#[uniffi::export]
+pub fn new_hd_path_component_from_global_key_space(
+    value: u32,
+) -> HDPathComponent {
+    InternalHDPathComponent::from_global_key_space(value)
+        .into_result()
+        .expect("Should always be able to create an HDPathComponent from a u32")
+}
+
+#[uniffi::export]
+pub fn hd_path_component_to_bip32_string(component: HDPathComponent) -> String {
+    Into::<InternalHDPathComponent>::into(component).to_bip32_string()
```

Comment:
you can use `component.into_internal().to_bip32_string()`

Reply:
nice